lib/mongo/server/connection_pool/queue.rb

```diff
@@ -190,31 +216,21 @@ def disconnect!
         #
         # @since 2.0.0
         def enqueue(connection)
-          check_count_invariants
+          check_count_invariants(false)
           mutex.synchronize do
             if connection.generation == @generation
               queue.unshift(connection.record_checkin!)
               resource.broadcast
-            else
-              connection.disconnect!
-
-              @pool_size = if @pool_size > 0
-                @pool_size - 1
-              else
-                # This should never happen
-                log_warn("ConnectionPool::Queue: unexpected enqueue")
-                0
-              end
-
-              while @pool_size < min_size
-                @pool_size += 1
-                queue.unshift(@block.call(@generation))
+              @wait_queue_mutex.synchronize do
+                @wait_queue.first.broadcast unless @wait_queue.empty?
```

Comment:
Why are there two broadcast calls in this method?

Reply:
Because there are two separate semaphores being signaled.

Reply:
I believe this has been resolved by our offline discussions and the recent updates